edward/inferences/renyi_divergence.py

```diff
@@ -0,0 +1,191 @@
+        """Build the Renyi ELBO function.
+
+        Its automatic differentiation is a stochastic gradient of
+
+        $ \mcalL_{R}^{\alpha}(q; x) =
+                \frac{1}{1-\alpha} \log \dsE_{q} \left[
+                    \left( \frac{p(x, z)}{q(z)}\right)^{1-\alpha} \right] $
+
+        It uses:
+        1. Monte Carlo approximation of the ELBO (Y. Li & al, 2016)
+        2. Reparameterization gradients (Kingma & al, 2014)
+        3. Stochastic approximation of the joint distribution (Y. Li & al, 2016)
+
+        # Notes
+            If the model is not reparameterizable, it returns a
+            NotImplementedError.
+            See Renyi Divergence Variational Inference (Y. Li & al, 2016)
+            for more details.
+        """
+        is_reparameterizable = all([
+            rv.reparameterization_type ==
+            tf.contrib.distributions.FULLY_REPARAMETERIZED
+            for rv in six.itervalues(self.latent_vars)])
+
+        if is_reparameterizable:
+            p_log_prob = [0.0] * self.n_samples
+            q_log_prob = [0.0] * self.n_samples
+            base_scope = tf.get_default_graph().unique_name("inference") + '/'
+            for s in range(self.n_samples):
+                # Form dictionary in order to replace conditioning on prior or
+                # observed variable with conditioning on a specific value.
+                scope = base_scope \
+                    + tf.get_default_graph().unique_name("sample")
+                dict_swap = {}
+                for x, qx in six.iteritems(self.data):
+                    if isinstance(x, RandomVariable):
+                        if isinstance(qx, RandomVariable):
+                            qx_copy = copy(qx, scope=scope)
+                            dict_swap[x] = qx_copy.value()
+                        else:
+                            dict_swap[x] = qx
+
+                for z, qz in six.iteritems(self.latent_vars):
+                    # Copy q(z) to obtain new set of posterior samples.
+                    qz_copy = copy(qz, scope=scope)
+                    dict_swap[z] = qz_copy.value()
+                    q_log_prob[s] += tf.reduce_sum(
+                        self.scale.get(z, 1.0)
+                        * qz_copy.log_prob(dict_swap[z]))
+
+                for z in six.iterkeys(self.latent_vars):
+                    z_copy = copy(z, dict_swap, scope=scope)
+                    p_log_prob[s] += tf.reduce_sum(
+                        self.scale.get(z, 1.0)
+                        * z_copy.log_prob(dict_swap[z]))
+
+                for x in six.iterkeys(self.data):
+                    if isinstance(x, RandomVariable):
+                        x_copy = copy(x, dict_swap, scope=scope)
+                        p_log_prob[s] += tf.reduce_sum(
+                            self.scale.get(x, 1.0)
+                            * x_copy.log_prob(dict_swap[x]))
+
+            logF = [p - q for p, q in zip(p_log_prob, q_log_prob)]
```

Comment:
Instead of `logF`, what about something like `log_ratios`, which is more Pythonic in snake_case and also more semantically meaningful?